### PR TITLE
Add opensearch package and module

### PR DESCRIPTION
**Pull request by dpausp**
###### Description of changes

WIP: Add package and module for https://opensearch.org (#151464). Packaging with inspiration from #175000. Module is basically a copy of the elasticsearch module. Module works for me, but more testing required.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by shyim** [comment]
Hey @dpausp ,

looks good. Can you add a test for this module? if the setup basically works?

File: nixos/modules/services/search/opensearch.nix [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+      type = types.str;
+    };
+
+    single_node = mkOption {
+      description = lib.mdDoc "Start a single-node cluster";
+      default = true;
+      type = types.bool;
+    };
+
+    extraConf = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/search/opensearch.nix:78]: Maybe just replace this with a `settings` option? `freeform` options? See https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md for details.

File: nixos/modules/services/search/opensearch.nix [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+      type = types.listOf types.str;
+    };
+
+    extraJavaOptions = mkOption {
+      description = lib.mdDoc "Extra command line options for Java.";
+      default = [ ];
+      type = types.listOf types.str;
+      example = [ "-Djava.net.preferIPv4Stack=true" ];
+    };
+
```
Comment (aanderse) [inline nixos/modules/services/search/opensearch.nix:126]: Please keep whitespace somewhat consistent.

**Review by aanderse — COMMENTED** [review]
I left a couple comments on the module portion of this PR. I hope my comments help. Feel free to ping if you have any questions or need help.

**Comment by dpausp** [comment]
@aanderse most of the code issues are just copied over from the elasticsearch module because I wanted to have a working solution quickly but I'm working on it, thanks for the helpful comments :)

**Comment by dpausp** [comment]
@shyim yes, NixOS test is WIP

**Comment by SuperSandro2000** [comment]
> @aanderse most of the code issues are just copied over from the elasticsearch module because I wanted to have a working solution quickly but I'm working on it, thanks for the helpful comments :)

If most things are copied over then does an extra module even make sense?

**Comment by SuperSandro2000** [comment]
Closing in favor of the more active #175000

**Pull request closed without merging**